File: ios/SiezaGym/DesignSystem/Pantallas.swift
import ImageIO
import SwiftUI
import UIKit

/// El marco de una pantalla: rótulo opcional arriba, título, y una acción a la
/// derecha. Es el `PageShell` de la web.
struct Pantalla<Contenido: View, Accion: View>: View {
    @Environment(\.tema) private var tema
    @Environment(\.dismiss) private var dismiss

    let titulo: String
    var rotulo: String?
    /// Con `volver` aparece la flecha a la izquierda.
    var volver = false
    @ViewBuilder var accion: Accion
    @ViewBuilder var contenido: Contenido

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                HStack(alignment: .center, spacing: 12) {
                    if volver {
                        Button { dismiss() } label: {
                            Image(systemName: "arrow.left")
                                .font(.system(size: 17, weight: .medium))
                                .foregroundStyle(tema.texto)
                                .frame(width: 44, height: 44)
                        }
                        .accessibilityLabel("Volver")
                    }

                    VStack(alignment: .leading, spacing: 4) {
                        if let rotulo {
                            Text(rotulo)
                                .font(.system(size: 13))
                                .foregroundStyle(tema.texto2)
                        }
                        Text(titulo)
                            .font(.system(size: volver ? 24 : 30, weight: tema.plano ? .bold : .heavy))
                            .tracking(-0.7)
                            .foregroundStyle(tema.texto)
                            .lineLimit(2)
                            .minimumScaleFactor(0.7)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)

                    accion
                }
                .padding(.bottom, 4)

                contenido
            }
            .padding(.horizontal, 18)
            .padding(.top, 8)
            .padding(.bottom, 24)
        }
        .background { Backdrop() }
        .scrollIndicators(.hidden)
        .scrollDismissesKeyboard(.interactively)
        .tecladoConBotonListo()
        .navigationBarBackButtonHidden(volver)
        .toolbar(.hidden, for: .navigationBar)
    }
}

extension Pantalla where Accion == EmptyView {
    init(titulo: String, rotulo: String? = nil, volver: Bool = false, @ViewBuilder contenido: () -> Contenido) {
        self.init(titulo: titulo, rotulo: rotulo, volver: volver, accion: { EmptyView() }, contenido: contenido)
    }
}

/// Tres números en una tarjeta, separados por líneas. El `d2-stats` de la web.
struct StatsCard: View {
    @Environment(\.tema) private var tema
    let datos: [(valor: String, rotulo: String)]

    var body: some View {
        HStack(spacing: 0) {
            ForEach(Array(datos.enumerated()), id: \.offset) { indice, dato in
                if indice > 0 {
                    Rectangle().fill(tema.borde).frame(width: 1, height: 34)
                }
                VStack(spacing: 4) {
                    Text(dato.valor)
                        .font(.system(size: 22, weight: .semibold))
                        .monospacedDigit()
                        .tracking(-0.7)
                        .foregroundStyle(tema.texto)
                        .lineLimit(1)
                        .minimumScaleFactor(0.6)
                    Text(dato.rotulo)
                        .font(.system(size: 10))
                        .foregroundStyle(tema.texto2)
                        .multilineTextAlignment(.center)
                }
                .frame(maxWidth: .infinity)
            }
        }
        .padding(.vertical, 16)
        .padding(.horizontal, 10)
        .background(tema.vidrio(1), in: .rect(cornerRadius: tema.plano ? 18 : 26))
        .overlay {
            RoundedRectangle(cornerRadius: tema.plano ? 18 : 26)
                .strokeBorder(tema.borde, lineWidth: 1)
        }
    }
}

/// Panel de vidrio con filas separadas por una línea fina, como
/// `.d2-routine-list`: la lista se lee como un objeto y no como una pila.
struct PanelLista<Contenido: View>: View {
    @Environment(\.tema) private var tema
    @ViewBuilder var contenido: Contenido

    var body: some View {
        VStack(spacing: 0) { contenido }
            .background(tema.vidrio(1), in: .rect(cornerRadius: tema.plano ? 18 : 24))
            .overlay {
                RoundedRectangle(cornerRadius: tema.plano ? 18 : 24)
                    .strokeBorder(tema.borde, lineWidth: 1)
            }
    }
}

/// Una fila de lista: nombre, detalle abajo, y un valor opcional a la derecha.
struct FilaLista: View {
    @Environment(\.tema) private var tema
    let nombre: String
    let detalle: String
    var etiqueta: String?
    var valor: String?
    var unidad: String?
    var miniatura: URL?
    var chevron = true

    var body: some View {
        HStack(spacing: 12) {
            if let miniatura {
                Miniatura(url: miniatura, lado: 40)
            }

            VStack(alignment: .leading, spacing: 5) {
                HStack(spacing: 8) {
                    Text(nombre)
                        .font(.system(size: 15, weight: .medium))
                        .tracking(-0.2)
                        .foregroundStyle(tema.texto)
                        .lineLimit(1)
                    if let etiqueta {
                        Text(etiqueta)
                            .font(.system(size: 10))
                            .foregroundStyle(tema.texto2)
                    }
                }
                Text(detalle)
                    .font(.system(size: 11))
                    .foregroundStyle(tema.texto2)
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            if let valor {
                VStack(alignment: .trailing, spacing: 3) {
                    Text(valor)
                        .font(.system(size: 13))
                        .foregroundStyle(tema.texto)
                    if let unidad {
                        Text(unidad)
                            .font(.system(size: 9))
                            .foregroundStyle(tema.texto3)
                    }
                }
            }

            if chevron {
                Image(systemName: "chevron.right")
                    .font(.system(size: 13, weight: .medium))
                    .foregroundStyle(tema.texto3)
            }
        }
        .padding(.horizontal, 18)
        .padding(.vertical, 13)
        .contentShape(.rect)
    }
}

/// La animación del ejercicio. Fondo claro siempre: son trazos negros sobre
/// blanco y sobre el vidrio de un tema oscuro no se ven.
struct Miniatura: View {
    @Environment(\.tema) private var tema
    let url: URL?
    var lado: CGFloat = 54

    var body: some View {
        ZStack {
            Color(white: 0.95)
            if let url {
                if url.pathExtension.lowercased() == "gif" {
                    GIFMiniatura(url: url)
                } else {
                    AsyncImage(url: url) { imagen in
                        imagen.resizable().aspectRatio(contentMode: .fill)
                    } placeholder: {
                        PlaceholderMiniatura(lado: lado)
                    }
                }
            } else {
                PlaceholderMiniatura(lado: lado)
            }
        }
        .frame(width: lado, height: lado)
        .clipShape(.rect(cornerRadius: lado * 0.3))
    }
}

private struct PlaceholderMiniatura: View {
    let lado: CGFloat

    var body: some View {
        Image(systemName: "dumbbell.fill")
            .font(.system(size: lado * 0.34))
            .foregroundStyle(Color(white: 0.6))
    }
}

/// `AsyncImage` no reproduce GIFs remotos de forma fiable en iOS. Este camino
/// decodifica sus frames con ImageIO y los entrega a UIImageView, que sí los
/// anima. La caché evita descargar el mismo ejercicio en cada fila/pantalla.
private struct GIFMiniatura: UIViewRepresentable {
    let url: URL

    func makeCoordinator() -> Coordinator { Coordinator() }

    func makeUIView(context: Context) -> UIImageView {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFill
        imageView.clipsToBounds = true
        context.coordinator.load(url: url, into: imageView)
        return imageView
    }

    func updateUIView(_ imageView: UIImageView, context: Context) {
        context.coordinator.load(url: url, into: imageView)
    }

    static func dismantleUIView(_ imageView: UIImageView, coordinator: Coordinator) {
        coordinator.cancel()
    }

    @MainActor
    final class Coordinator {
        private var task: Task<Void, Never>?
        private var loadedURL: URL?

        func load(url: URL, into imageView: UIImageView) {
            guard loadedURL != url else { return }
            loadedURL = url
            task?.cancel()
            imageView.stopAnimating()
            imageView.animationImages = nil
            imageView.image = UIImage(systemName: "dumbbell.fill")

            task = Task { @MainActor [weak self, weak imageView] in
                guard let decoded = try? await GIFImageLoader.load(url: url),
                      !Task.isCancelled,
                      let self,
                      let imageView else { return }

                imageView.animationImages = decoded.frames
                imageView.animationDuration = decoded.duration
                imageView.animationRepeatCount = 0
                imageView.startAnimating()
                self.task = nil
            }
        }

        func cancel() {
            task?.cancel()
            task = nil
        }
    }
}

@MainActor
private enum GIFImageLoader {
    struct DecodedImage {
        let frames: [UIImage]
        let duration: TimeInterval
    }

    static let cache = NSCache<NSURL, UIImage>()

    static func load(url: URL) async throws -> DecodedImage {
        if let cached = cache.object(forKey: url as NSURL),
           let frames = cached.images,
           !frames.isEmpty {
            return DecodedImage(frames: frames, duration: max(cached.duration, 0.1))
        }

        let (data, _) = try await URLSession.shared.data(from: url)
        guard let source = CGImageSourceCreateWithData(data as CFData, nil) else {
            throw GIFImageError.invalidData
        }

        let count = CGImageSourceGetCount(source)
        guard count > 0 else { throw GIFImageError.invalidData }

        var frames: [UIImage] = []
        var duration: TimeInterval = 0
        for index in 0..<count {
            guard let image = CGImageSourceCreateImageAtIndex(source, index, nil) else { continue }
            frames.append(UIImage(cgImage: image))
            duration += frameDuration(source: source, index: index)
        }

        guard !frames.isEmpty else { throw GIFImageError.invalidData }
        let totalDuration = max(duration, Double(frames.count) * 0.1)
        let animated = UIImage.animatedImage(with: frames, duration: totalDuration) ?? frames[0]
        cache.setObject(animated, forKey: url as NSURL)
        return DecodedImage(frames: animated.images ?? frames, duration: totalDuration)
    }

    private static func frameDuration(source: CGImageSource, index: Int) -> TimeInterval {
        let properties = CGImageSourceCopyPropertiesAtIndex(source, index, nil) as? [CFString: Any]
        let gif = properties?[kCGImagePropertyGIFDictionary] as? [CFString: Any]
        let unclamped = gif?[kCGImagePropertyGIFUnclampedDelayTime] as? Double
        let clamped = gif?[kCGImagePropertyGIFDelayTime] as? Double
        return max(unclamped ?? clamped ?? 0.1, 0.02)
    }

    private enum GIFImageError: Error {
        case invalidData
    }
}

/// Estado vacío: qué falta y cómo salir de ahí.
struct Vacio: View {
    @Environment(\.tema) private var tema
    let texto: String
    var accion: (titulo: String, hacer: () -> Void)?

    var body: some View {
        VStack(spacing: 16) {
            Text(texto)
                .font(.system(size: 14))
                .foregroundStyle(tema.texto2)
                .multilineTextAlignment(.center)
            if let accion {
                Button(accion.titulo, action: accion.hacer)
                    .buttonStyle(SolidButtonStyle(expands: false))
            }
        }
        .frame(maxWidth: .infinity)
        .padding(28)
        .background(tema.vidrio(1), in: .rect(cornerRadius: 24))
        .overlay { RoundedRectangle(cornerRadius: 24).strokeBorder(tema.borde, lineWidth: 1) }
    }
}

/// La atribución de los gifs. Es condición de la licencia: donde se ven las
/// animaciones tiene que estar el crédito.
struct CreditoGifs: View {
    @Environment(\.tema) private var tema

    var body: some View {
        Link(destination: URL(string: "https://gymvisual.com/")!) {
            Text("Animaciones de ejercicios © Gym visual")
                .font(.system(size: 10))
                .underline()
                .foregroundStyle(tema.texto3)
        }
        .frame(maxWidth: .infinity)
        .padding(.top, 14)
    }
}

File: android/app/src/main/kotlin/com/siezagym/app/DesignSystem/Pantallas.kt
package com.siezagym.app.DesignSystem

import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.FitnessCenter
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil3.compose.AsyncImage

/** El marco de una pantalla: rótulo opcional arriba, título, y una acción a la
 * derecha. Es el `PageShell` de la web. */
@Composable
fun Pantalla(
    titulo: String,
    rotulo: String? = null,
    volver: Boolean = false,
    onVolver: () -> Unit = {},
    accion: @Composable RowScope.() -> Unit = {},
    contenido: @Composable ColumnScope.() -> Unit,
) {
    val tema = LocalD2Theme.current
    Box(Modifier.fillMaxSize()) {
        Backdrop()
        Column(
            Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 18.dp)
                .padding(top = 8.dp, bottom = 24.dp),
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                if (volver) {
                    IconButton(
                        onClick = onVolver,
                        modifier = Modifier.size(44.dp),
                    ) {
                        Icon(
                            Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = "Volver",
                            tint = tema.texto,
                        )
                    }
                }
                Column(Modifier.weight(1f)) {
                    if (rotulo != null) {
                        Text(rotulo, fontSize = 13.sp, color = tema.texto2)
                    }
                    Text(
                        titulo,
                        fontSize = if (volver) 24.sp else 30.sp,
                        fontWeight = FontWeight.Heavy,
                        letterSpacing = (-0.7).sp,
                        color = tema.texto,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
                accion()
            }
            Spacer(Modifier.height(4.dp))
            contenido()
        }
    }
}

/** Tres números en una tarjeta, separados por líneas. El `d2-stats` de la web. */
@Composable
fun StatsCard(datos: List<Pair<String, String>>, modifier: Modifier = Modifier) {
    val tema = LocalD2Theme.current
    Row(modifier = modifier.fillMaxWidth()) {
        datos.forEachIndexed { indice, dato ->
            if (indice > 0) {
                Box(
                    Modifier
                        .height(34.dp)
                        .width(1.dp)
                        .background(tema.borde),
                )
            }
            Column(
                Modifier.weight(1f),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Spacer(Modifier.height(16.dp))
                Text(
                    dato.first,
                    fontSize = 22.sp,
                    fontWeight = FontWeight.SemiBold,
                    letterSpacing = (-0.7).sp,
                    color = tema.texto,
                    maxLines = 1,
                )
                Spacer(Modifier.height(4.dp))
                Text(
                    dato.second,
                    fontSize = 10.sp,
                    color = tema.texto2,
                    textAlign = TextAlign.Center,
                )
            }
        }
        Spacer(Modifier.height(16.dp))
    }
}

/** Panel de vidrio con filas separadas por una línea fina, como
 * `.d2-routine-list`: la lista se lee como un objeto y no como una pila. */
@Composable
fun PanelLista(
    modifier: Modifier = Modifier,
    contenido: @Composable ColumnScope.() -> Unit,
) {
    val tema = LocalD2Theme.current
    Column(
        modifier
            .clip(RoundedCornerShape(24.dp))
            .background(tema.vidrio(1))
            .border(1.dp, tema.borde, RoundedCornerShape(24.dp)),
        content = contenido,
    )
}

/** Una fila de lista: nombre, detalle abajo, y un valor opcional a la derecha. */
@Composable
fun FilaLista(
    nombre: String,
    detalle: String,
    modifier: Modifier = Modifier,
    etiqueta: String? = null,
    valor: String? = null,
    unidad: String? = null,
    miniatura: String? = null,
    chevron: Boolean = true,
    onClick: (() -> Unit)? = null,
) {
    val tema = LocalD2Theme.current
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = modifier
            .fillMaxWidth()
            .clickable(enabled = onClick != null, onClick = { onClick?.invoke() })
            .padding(horizontal = 18.dp, vertical = 13.dp),
    ) {
        if (miniatura != null) {
            Miniatura(url = miniatura, lado = 40.dp)
            Spacer(Modifier.width(12.dp))
        }

        Column(Modifier.weight(1f)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    nombre,
                    fontSize = 15.sp,
                    fontWeight = FontWeight.Medium,
                    letterSpacing = (-0.2).sp,
                    color = tema.texto,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                if (etiqueta != null) {
                    Spacer(Modifier.width(8.dp))
                    Text(etiqueta, fontSize = 10.sp, color = tema.texto2)
                }
            }
            Spacer(Modifier.height(5.dp))
            Text(
                detalle,
                fontSize = 11.sp,
                color = tema.texto2,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
        }

        if (valor != null) {
            Spacer(Modifier.width(12.dp))
            Column(horizontalAlignment = Alignment.End) {
                Text(
                    valor,
                    fontSize = 13.sp,
                    color = tema.texto,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                if (unidad != null) {
                    Text(unidad, fontSize = 9.sp, color = tema.texto3)
                }
            }
        }

        if (chevron) {
            Spacer(Modifier.width(12.dp))
            Icon(
                Icons.Filled.ChevronRight,
                contentDescription = null,
                tint = tema.texto3,
                modifier = Modifier.size(16.dp),
            )
        }
    }
}

/** La animación del ejercicio. Fondo claro siempre: son trazos negros sobre
 * blanco y sobre el vidrio de un tema oscuro no se ven.
 *
 * Coil ya viene con el decodificador de GIF, así que los gifs remotos animan
 * solos. */
@Composable
fun Miniatura(url: String?, lado: Dp = 54.dp, modifier: Modifier = Modifier) {
    Box(
        modifier
            .size(lado)
            .clip(RoundedCornerShape(lado * 0.3f))
            .background(Color(0xFFF2F2F2)),
        contentAlignment = Alignment.Center,
    ) {
        if (!url.isNullOrBlank()) {
            AsyncImage(
                model = url,
                contentDescription = null,
                contentScale = ContentScale.Crop,
                modifier = Modifier.fillMaxSize(),
            )
        }
        Icon(
            Icons.Filled.FitnessCenter,
            contentDescription = null,
            tint = Color(0xFF999999),
            modifier = Modifier.size(lado * 0.38f),
        )
    }
}

/** Estado vacío: qué falta y cómo salir de ahí. */
@Composable
fun Vacio(
    texto: String,
    modifier: Modifier = Modifier,
    accionTitulo: String? = null,
    onAccion: (() -> Unit)? = null,
) {
    val tema = LocalD2Theme.current
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = modifier
            .fillMaxWidth()
            .padding(28.dp)
            .clip(RoundedCornerShape(24.dp))
            .background(tema.vidrio(1))
            .border(1.dp, tema.borde, RoundedCornerShape(24.dp)),
    ) {
        Spacer(Modifier.height(16.dp))
        Text(
            texto,
            fontSize = 14.sp,
            color = tema.texto2,
            textAlign = TextAlign.Center,
        )
        if (accionTitulo != null && onAccion != null) {
            Spacer(Modifier.height(16.dp))
            GhostButton(texto = accionTitulo, onClick = onAccion)
        }
        Spacer(Modifier.height(16.dp))
    }
}

/** La atribución de los gifs. Es condición de la licencia: donde se ven las
 * animaciones tiene que estar el crédito. */
@Composable
fun CreditoGifs(modifier: Modifier = Modifier) {
    val tema = LocalD2Theme.current
    val contexto = LocalContext.current
    Text(
        "Animaciones de ejercicios © Gym visual",
        fontSize = 10.sp,
        color = tema.texto3,
        textDecoration = androidx.compose.ui.text.style.TextDecoration.Underline,
        textAlign = TextAlign.Center,
        modifier = modifier
            .fillMaxWidth()
            .padding(top = 14.dp)
            .clickable {
                contexto.startActivity(
                    Intent(Intent.ACTION_VIEW, Uri.parse("https://gymvisual.com/")),
                )
            },
    )
}